src/main/scala/pbdirect/PBMessageWriter.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2019 Beyond the lines
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package pbdirect
+
+import com.google.protobuf.CodedOutputStream
+import shapeless._
+import shapeless.ops.hlist._
+
+trait PBMessageWriter[A] {
+  def writeTo(value: A, out: CodedOutputStream): Unit
+}
+
+trait PBMessageWriterImplicits {
+  def instance[A](f: (A, CodedOutputStream) => Unit): PBMessageWriter[A] =
+    new PBMessageWriter[A] {
+      override def writeTo(value: A, out: CodedOutputStream): Unit =
+        f(value, out)
+    }
+  implicit val hnilWriter: PBMessageWriter[HNil] = instance { (_: HNil, _: CodedOutputStream) =>
+    ()
+  }
+  implicit def consWriter[H, T <: HList](
+      implicit head: PBFieldWriter[H],
+      tail: Lazy[PBMessageWriter[T]]): PBMessageWriter[(FieldIndex, H) :: T] =
+    instance { (value: (FieldIndex, H) :: T, out: CodedOutputStream) =>
+      val headIndex = value.head._1.values.head
+      val headValue = value.head._2
+      head.writeTo(headIndex, headValue, out)
+      tail.value.writeTo(value.tail, out)
+    }
+
+  object zipWithFieldIndex extends Poly2 {
+    implicit def defaultCase[T] = at[Some[pbIndex], T] {
+      case (Some(annotation), value) =>
+        (FieldIndex(annotation.first :: annotation.more.toList), value)
+    }
+  }
+
```

Comment:
This is probably the most interesting part of the PR.

Given some case class instance, e.g.

```scala
case class MyMessage(@pbIndex(1) a: Int, @pbIndex(2) b: String)
val msg = MyMessage(123, "hello")
```

we summon the generic representation:

```scala
123 :: "hello" :: HNil
```

and the list of `@pbIndex` annotations on the fields:

```scala
Some(pbIndex(1)) :: Some(pbIndex(2)) :: HNil
```

We validate that every field is annotated and zip the two `HList`s together using the `zipWithFieldIndex` polymorphic function, giving us an `HList` of tuples:

```scala
(FieldIndex(1), 123) :: (FieldIndex(2), "hello") :: HNil
```

Then we use `consWriter` and `hnilWriter` to inductively resolve the correct `PBFieldWriter` instance for each field and write the fields to the output stream.